### PR TITLE
Do not fire a query when changing options if unsubscribed

### PR DESCRIPTION
**Pull request by tmeasday**
WIP - failing test for https://github.com/apollostack/apollo-client/issues/1077

TODO:

- [ ] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [ ] Make sure all tests and linter rules pass
- [ ] Update CHANGELOG.md with your change
- [ ] Add your name and email to the AUTHORS file (optional)

### CODE REVIEW

**Comment by tmeasday** [comment]
@helfer any thoughts about how to best fix this? The issue is that `setVariables/Options` return a "result" promise and this kind of forces us to make a request.

The motivation for that request was
a) to update any subscribers
b) to give the caller a promise so they knew when the "updating" was done.

In the the case that there are no subscribes, (a) is not required, so therefore (b) is instantaneous. 

We could change the API so that it returns an optional promise, or make the promise return `null` (we don't know the result of the query) straightaway in this case.

What do you think?

**Comment by helfer** [comment]
@tmeasday Yeah, good point. I'd be okay with returning `null`. Depending on how `setVariables` is used, it might also make sense to reject the promise to say that there's no new result, but that might get confusing. Let's go with `null`.

I'm curious though, what's the scenario that made the error show up for you?

**Comment by tmeasday** [comment]
When you say "returning `null`" do you mean a promise that returns `null`?

So I think you found the original problem (https://github.com/apollostack/react-apollo/blob/7eae575d4cf66b6c934fd34ce0db4f337fd6800d/test/react-web/client/graphql/queries.test.tsx#L523-L524) but for the record, the issue is when you:

1. Set up a query with RA
2. (Later) Skip the query
3. (Later) Unskip the query at the same time as changing the variables.

Step 3 triggers two network requests because changing the variables on the query triggers a request, even though it has no subscribers; then setting up the first subscriber does the same.

**Comment by helfer** [comment]
@tmeasday yes, I meant a promise that resolves to `null`. 🙂

File: src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -304,19 +304,41 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
...
+      // If we have no observers, then we don't actually want to make a network
+      // request. As soon as someone observes the query, the request will kick
+      // off. For now, we just store any changes. (See #1077)
+      if (this.observers.length === 0) {
+        return nullPromise;
+      }
+
       return this.result();
     } else {
       this.variables = newVariables;
```
Comment (tmeasday) [inline src/core/ObservableQuery.ts]: @helfer I don't know why we don't set `this.options.variables` too here, but it seems weird. Why do we even have `this.variables`? I don't recall a good reason. Do you know?

**Comment by tmeasday** [comment]
@helfer I think this change is good but it's a small refactor so you might want to think about it a bit.

**Comment by tmeasday** [comment]
ping @helfer

File: src/core/ObservableQuery.ts (hunk 1 above)
Comment (helfer) [inline src/core/ObservableQuery.ts]: I don't know, and I think there are / were some bugs related to that. We should just refactor to remove this.variables.

Comment (tmeasday) [inline src/core/ObservableQuery.ts]: OK, I'll do it in this PR then?

**Comment by helfer** [comment]
Thanks Tom! I think the minor change is perfectly acceptable to fix the issue of fetching when no fetch is necessary.

**Comment by helfer** [comment]
Aand I forgot to add an entry to the changelog. That's why one should never get up in the middle of a task.

**Pull request merged**